File: misc/productexceptself/prefixsuffix/ProductExceptSelf.java
package productexceptself.prefixsuffix;

class ProductExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];

        // Step 1: Compute the left products
        result[0] = 1;
        for (int i = 1; i < n; i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }

        // Step 2: Compute the right products and update the result
        int rightProduct = 1;
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= rightProduct;
            rightProduct *= nums[i];
        }

        return result;
    }

    // Sample usage
    public static void main(String[] args) {
        ProductExceptSelf solution = new ProductExceptSelf();

        // Test case 1
        int[] nums1 = {1, 2, 3, 4};
        int[] result1 = solution.productExceptSelf(nums1);
        for (int num : result1) {
            System.out.print(num + " ");  // Output: 24 12 8 6
        }
        System.out.println();

        // Test case 2
        int[] nums2 = {-1, 1, 0, -3, 3};
        int[] result2 = solution.productExceptSelf(nums2);
        for (int num : result2) {
            System.out.print(num + " ");  // Output: 0 0 9 0 0
        }
        System.out.println();
    }
}

File: misc/productexceptself/prefixsuffix/ProductExceptSelf.py
class ProductExceptSelf:
    def productExceptSelf(self, nums):
        n = len(nums)
        answer = [1] * n  # Initialize result array with 1

        # Step 1: Compute the left products
        left_product = 1
        for i in range(n):
            answer[i] = left_product  # Set current answer to the left product
            left_product *= nums[i]  # Update left product for next iteration

        # Step 2: Compute the right products and update the result
        right_product = 1
        for i in range(n - 1, -1, -1):
            answer[i] *= right_product  # Multiply with the right product
            right_product *= nums[i]  # Update right product for next iteration

        return answer

# Sample usage
if __name__ == "__main__":
    solution = ProductExceptSelf()

    # Test case 1
    nums1 = [1, 2, 3, 4]
    print(solution.productExceptSelf(nums1))  # Output: [24, 12, 8, 6]

    # Test case 2
    nums2 = [-1, 1, 0, -3, 3]
    print(solution.productExceptSelf(nums2))  # Output: [0, 0, 9, 0, 0]
